File: other_benchmarks/floyd_warshall/floyd_warshall.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define INF 99999 // A large value representing infinity

double diff_timespec(struct timespec *time1, struct timespec *time0)
{
    return (time1->tv_sec - time0->tv_sec) + (time1->tv_nsec - time0->tv_nsec) / 1000000000.0;
}

void printSolution(int **dist, int V)
{
    printf("Shortest distances between every pair of vertices:\n");
    for (int i = 0; i < V; i++)
    {
        for (int j = 0; j < V; j++)
        {
            if (dist[i][j] == INF)
                printf("INF ");
            else
                printf("%d ", dist[i][j]);
        }
        printf("\n");
    }
}

void floydWarshall(int **graph, int V)
{
    int **dist = (int **)malloc(V * sizeof(int *));
    for (int i = 0; i < V; i++)
    {
        dist[i] = (int *)malloc(V * sizeof(int));
        for (int j = 0; j < V; j++)
        {
            dist[i][j] = graph[i][j];
        }
    }

    for (int k = 0; k < V; k++)
    {
        for (int i = 0; i < V; i++)
        {
            for (int j = 0; j < V; j++)
            {
                if (dist[i][k] != INF && dist[k][j] != INF && dist[i][k] + dist[k][j] < dist[i][j])
                {
                    dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
    }

    // printSolution(dist, V);

    for (int i = 0; i < V; i++)
        free(dist[i]);
    free(dist);
}

int main(int argc, char *argv[])
{
    if (argc != 2)
    {
        printf("Usage: %s <number of vertices>\n", argv[0]);
        return 1;
    }

    int V = atoi(argv[1]);
    if (V <= 0)
    {
        printf("Number of vertices must be positive.\n");
        return 1;
    }
    struct timespec start_time, end_time;
    int **graph = (int **)malloc(V * sizeof(int *));

    srand(10000);
    for (int i = 0; i < V; i++)
    {
        graph[i] = (int *)malloc(V * sizeof(int));
        for (int j = 0; j < V; j++)
        {
            if (i == j)
            {
                graph[i][j] = 0;
            }
            else
            {
                if (rand() % 4 == 0) // Adjust probability of INF edges
                    graph[i][j] = INF;
                else
                    graph[i][j] = rand() % 20 + 1;
            }
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &start_time);

    floydWarshall(graph, V);
    clock_gettime(CLOCK_MONOTONIC, &end_time);

    double time_elapsed = diff_timespec(&end_time, &start_time);
    printf("Time taken to find shortest distances of graph size %d: %f seconds\n", V, time_elapsed);

    for (int i = 0; i < V; i++)
        free(graph[i]);
    free(graph);

    return 0;
}

File: other_benchmarks/floyd_warshall/src/floyd_warshall.rs
use libc::{clock_gettime, rand, srand, timespec, CLOCK_MONOTONIC};
use std::env;
use std::process::exit;

const INF: i32 = 99999; // Large value representing infinity

// Function to compute the difference between two timespec structs
fn diff_timespec(time1: &timespec, time0: &timespec) -> f64 {
    (time1.tv_sec - time0.tv_sec) as f64 + (time1.tv_nsec - time0.tv_nsec) as f64 / 1_000_000_000.0
}

// Function to fetch the current time using clock_gettime
fn get_time() -> timespec {
    let mut ts: timespec = unsafe { std::mem::zeroed() };
    unsafe { clock_gettime(CLOCK_MONOTONIC, &mut ts) };
    ts
}

// Floyd-Warshall Algorithm implementation
fn floyd_warshall(graph: &mut Vec<Vec<i32>>, v: usize) {
    for k in 0..v {
        for i in 0..v {
            for j in 0..v {
                if graph[i][k] != INF
                    && graph[k][j] != INF
                    && graph[i][k] + graph[k][j] < graph[i][j]
                {
                    graph[i][j] = graph[i][k] + graph[k][j];
                }
            }
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <number of vertices>", args[0]);
        exit(1);
    }

    let v: usize = match args[1].parse() {
        Ok(n) if n > 0 => n,
        _ => {
            eprintln!("Number of vertices must be positive.");
            exit(1);
        }
    };

    let mut graph = vec![vec![0; v]; v];

    unsafe { srand(10000) }; // Seed random number generator

    for i in 0..v {
        for j in 0..v {
            if i == j {
                graph[i][j] = 0;
            } else {
                let random_value = unsafe { rand() } % 4;
                if random_value == 0 {
                    graph[i][j] = INF;
                } else {
                    graph[i][j] = (unsafe { rand() } % 20 + 1) as i32;
                }
            }
        }
    }

    let start_time = get_time();
    floyd_warshall(&mut graph, v);
    let end_time = get_time();

    let time_elapsed = diff_timespec(&end_time, &start_time);
    println!(
        "Time taken to find shortest distances of graph size {}: {:.6} seconds",
        v, time_elapsed
    );
}
